crates/kornia-pnp/src/epnp.rs

```diff
@@ -0,0 +1,590 @@
+///
+/// # Returns
+/// Tuple *(R, t, rvec)* where
+/// * `R` – 3×3 rotation **world → camera**,
+/// * `t` – 3-vector translation,
+/// * `rvec` – Rodrigues axis-angle representation of `R`.
+pub fn solve_epnp(
+    world_pts: &[[f64; 3]],
+    image_pts: &[[f64; 2]],
+    k: &[[f64; 3]; 3],
+    params: &EPNPParams,
+) -> Result<PnPResult, &'static str> {
+    let n = world_pts.len();
+    if n != image_pts.len() || n < 4 {
+        return Err("EPnP requires ≥4 2D–3D correspondences");
+    }
+
+    let cw = select_control_points(world_pts);
+
+    let alphas = compute_barycentric(world_pts, &cw, params.tol.eps);
+
+    // Build the 2N×12 design matrix M
+    let m_rows = build_m(&alphas, image_pts, k);
+
+    let m_flat: Vec<f64> = m_rows.iter().flat_map(|row| row.iter()).cloned().collect();
+    let m_mat = DMatrix::<f64>::from_row_slice(2 * n, 12, &m_flat);
+
+    // Null-space of M (4 right-singular vectors associated with smallest singular values)
+    let svd = m_mat.svd(true, true);
+    let v_t = match svd.v_t {
+        Some(v) => v,
+        None => return Err("SVD failed to compute V^T"),
+    };
+    let cols = 12;
+    let start_col = cols - 4;
+    let null4 = v_t.rows(start_col, 4).transpose(); // shape 12×4
+
+    // Build helper matrices for beta initialisation
+    let l = build_l6x10(&null4);
+    let rho = rho_ctrlpts(&cw);
+
+    // Convert L and rho to nalgebra types once.
+    let rho_vec = DVector::<f64>::from_column_slice(&rho);
+
+    let mut betas: Vec<[f64; 4]> = Vec::new();
+
+    betas.extend(
+        [
+            estimate_beta([0, 1, 3, 6], &l, &rho_vec, params.tol.svd),
+            estimate_beta([0, 1, 2], &l, &rho_vec, params.tol.svd),
+            estimate_beta([0, 1, 2, 3, 4], &l, &rho_vec, params.tol.svd),
+        ]
+        .into_iter()
+        .flatten(),
+    );
+
+    let betas_refined: Vec<[f64; 4]> = betas
+        .iter()
+        .map(|&b| gauss_newton(b, &null4, &rho))
+        .collect();
+
+    let mut best_err = f64::INFINITY;
+    let mut best_r = [[1.0; 3]; 3];
+    let mut best_t = [0.0; 3];
+
+    for bet in &betas_refined {
+        let (r_c, t_c) = pose_from_betas(bet, &null4, &cw, &alphas);
+        let err = rmse_px(world_pts, image_pts, &r_c, &t_c, k);
+        if err < best_err {
+            best_err = err;
+            best_r = r_c;
+            best_t = t_c;
+        }
+    }
+
+    let mat = Mat3A::from_cols_array(&[
+        best_r[0][0] as f32,
+        best_r[1][0] as f32,
+        best_r[2][0] as f32,
+        best_r[0][1] as f32,
+        best_r[1][1] as f32,
+        best_r[2][1] as f32,
+        best_r[0][2] as f32,
+        best_r[1][2] as f32,
+        best_r[2][2] as f32,
+    ]);
+    let rvec_f32 = SO3::from_matrix(&mat).log();
+    let rvec = [rvec_f32.x as f64, rvec_f32.y as f64, rvec_f32.z as f64];
+
+    Ok(PnPResult {
+        rotation: best_r,
+        translation: best_t,
+        rvec,
+        reproj_rmse: Some(best_err),
```

Comment:
why are all these optionals ?

Reply:
This is as in opencv, can remove the rotation vector but the others are quite useful.